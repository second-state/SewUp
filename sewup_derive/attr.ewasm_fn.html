<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="helps you to build your handlers in the contract"><meta name="keywords" content="rust, rustlang, rust-lang, ewasm_fn"><title>ewasm_fn in sewup_derive - Rust</title><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../ayu.css" disabled ><script id="default-settings" ></script><script src="../storage.js"></script><script src="../crates.js"></script><script defer src="../main.js"></script>
    <noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"><style type="text/css">#crate-search{background-image:url("../down-arrow.svg");}</style></head><body class="rustdoc attr"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../sewup_derive/index.html'><div class='logo-container rust-logo'><img src='../rust-logo.png' alt='logo'></div></a><div class="sidebar-elems"><h2 class="location">Other items in<br><a href="index.html">sewup_derive</a></h2><div id="sidebar-vars" data-name="ewasm_fn" data-ty="attr" data-relpath=""></div><script defer src="sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="18" height="18" alt="Pick another theme!" src="../brush.svg"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../settings.html" title="settings"><img width="18" height="18" alt="Change settings" src="../wheel.svg"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Attribute Macro <a href="index.html">sewup_derive</a>::<wbr><a class="attr" href="#">ewasm_fn</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../src/sewup_derive/lib.rs.html#410" title="goto source code">[src]</a></span></h1><div class="docblock item-decl"><pre class="rust attr"><code>#[ewasm_fn]</code></pre></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>helps you to build your handlers in the contract</p>
<p>This macro also generate the function signature, you can use
<code>ewasm_fn_sig!</code> macro to get your function signature;</p>

<div class='information'><div class='tooltip compile_fail'>ⓘ</div></div><div class="example-wrap"><pre class="rust rust-example-rendered compile_fail"><code><span class="attribute">#[<span class="ident">ewasm_fn</span>]</span>
<span class="kw">fn</span> <span class="ident">check_input_object</span>(<span class="ident">s</span>: <span class="ident">SimpleStruct</span>) -&gt; <span class="ident">anyhow::Result</span><span class="op">&lt;</span>()<span class="op">&gt;</span> {
    <span class="prelude-val">Ok</span>(())
}

<span class="attribute">#[<span class="ident">ewasm_main</span>]</span>
<span class="kw">fn</span> <span class="ident">main</span>() -&gt; <span class="prelude-ty">Result</span><span class="op">&lt;</span>()<span class="op">&gt;</span> {
    <span class="kw">let</span> <span class="ident">contract</span> <span class="op">=</span> <span class="ident">Contract::new</span>()<span class="question-mark">?</span>;
    <span class="kw">match</span> <span class="ident">contract</span>.<span class="ident">get_function_selector</span>()<span class="question-mark">?</span> {
        <span class="macro">ewasm_fn_sig!</span>(<span class="ident">check_input_object</span>) =&gt; <span class="macro">ewasm_input_from!</span>(<span class="ident">contract</span> <span class="kw">move</span> <span class="ident">check_input_object</span>)<span class="question-mark">?</span>,
        <span class="kw">_</span> =&gt; <span class="kw">return</span> <span class="prelude-val">Err</span>(<span class="ident">Error::UnknownHandle</span>.<span class="ident">into</span>()),
    };
    <span class="prelude-val">Ok</span>(())
}</code></pre></div>
<p>There are two kind of inputs for <code>ewasm_fn</code> macro, first is functional signature, second and
more are the fields of abijson.
The functional signature can be specific as following ways.
<code>#[ewasm_fn(00fdd58e)]</code> or #[ewasm_fn(“00fdd58e”)]</p>
<div class="example-wrap"><pre class="language-#[ewasm_fn(00fdd58e,"><code>constant=true,
inputs=[
    { &quot;internalType&quot;: &quot;address&quot;, &quot;name&quot;: &quot;account&quot;, &quot;type&quot;: &quot;address&quot; },
    { &quot;internalType&quot;: &quot;uinit256&quot;, &quot;name&quot;: &quot;token_id&quot;, &quot;type&quot;: &quot;uinit256&quot; }
],
name=balanceOf,
outputs=[
    { &quot;internalType&quot;: &quot;uint256&quot;, &quot;name&quot;: &quot;&quot;, &quot;type&quot;: &quot;uint256&quot; }
],
payable=false,
stateMutability=view
)]</code></pre></div>
<p>The fields are not required, it can use default value if not provided.
The default values of <code>constant</code>, <code>payable</code> are <code>false</code>; the default values of <code>inputs</code> and
<code>outputs</code> are <code>[]</code>; the default value of <code>stateMutability</code> is <code>view</code>; the default name is the
camel case style of the function name.</p>
</div></details></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../" data-current-crate="sewup_derive" data-search-index-js="../search-index.js" data-search-js="../search.js"></div>
</body></html>