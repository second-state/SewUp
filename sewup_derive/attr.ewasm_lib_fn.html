<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="helps you to build your handler in other module"><meta name="keywords" content="rust, rustlang, rust-lang, ewasm_lib_fn"><title>ewasm_lib_fn in sewup_derive - Rust</title><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../ayu.css" disabled ><script id="default-settings" ></script><script src="../storage.js"></script><script src="../crates.js"></script><script defer src="../main.js"></script>
    <noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"><style type="text/css">#crate-search{background-image:url("../down-arrow.svg");}</style></head><body class="rustdoc attr"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../sewup_derive/index.html'><div class='logo-container rust-logo'><img src='../rust-logo.png' alt='logo'></div></a><div class="sidebar-elems"><h2 class="location">Other items in<br><a href="index.html">sewup_derive</a></h2><div id="sidebar-vars" data-name="ewasm_lib_fn" data-ty="attr" data-relpath=""></div><script defer src="sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="18" height="18" alt="Pick another theme!" src="../brush.svg"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../settings.html" title="settings"><img width="18" height="18" alt="Change settings" src="../wheel.svg"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Attribute Macro <a href="index.html">sewup_derive</a>::<wbr><a class="attr" href="#">ewasm_lib_fn</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../src/sewup_derive/lib.rs.html#393" title="goto source code">[src]</a></span></h1><div class="docblock item-decl"><pre class="rust attr"><code>#[ewasm_lib_fn]</code></pre></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>helps you to build your handler in other module</p>
<p>This macro will automatically generated as <code>{FUNCTION_NAME}_SIG</code></p>

<div class='information'><div class='tooltip compile_fail'>ⓘ</div></div><div class="example-wrap"><pre class="rust rust-example-rendered compile_fail"><code><span class="comment">// module.rs</span>

<span class="kw">use</span> <span class="ident">sewup::ewasm_api</span>;

<span class="attribute">#[<span class="ident">ewasm_lib_fn</span>]</span>
<span class="kw">pub</span> <span class="kw">fn</span> <span class="ident">symbol</span>(<span class="ident">s</span>: <span class="kw-2">&amp;</span><span class="ident">str</span>) {
    <span class="kw">let</span> <span class="ident">symbol</span> <span class="op">=</span> <span class="ident">s</span>.<span class="ident">to_string</span>().<span class="ident">into_bytes</span>();
    <span class="ident">ewasm_api::finish_data</span>(<span class="kw-2">&amp;</span><span class="ident">symbol</span>);
}</code></pre></div>

<div class='information'><div class='tooltip compile_fail'>ⓘ</div></div><div class="example-wrap"><pre class="rust rust-example-rendered compile_fail"><code><span class="comment">// lib.rs</span>

<span class="kw">use</span> <span class="ident">module</span>::{<span class="ident">symbol</span>, <span class="ident">SYMBOL_SIG</span>};

<span class="attribute">#[<span class="ident">ewasm_main</span>]</span>
<span class="kw">fn</span> <span class="ident">main</span>() -&gt; <span class="prelude-ty">Result</span><span class="op">&lt;</span>()<span class="op">&gt;</span> {
    <span class="kw">let</span> <span class="ident">contract</span> <span class="op">=</span> <span class="ident">Contract::new</span>()<span class="question-mark">?</span>;
    <span class="kw">match</span> <span class="ident">contract</span>.<span class="ident">get_function_selector</span>()<span class="question-mark">?</span> {
        <span class="ident">SYMBOL_SIG</span> =&gt; <span class="ident">symbol</span>(<span class="string">&quot;ETD&quot;</span>),
        <span class="kw">_</span> =&gt; <span class="kw">return</span> <span class="prelude-val">Err</span>(<span class="ident">Error::UnknownHandle</span>.<span class="ident">into</span>()),
    };
    <span class="prelude-val">Ok</span>(())
}</code></pre></div>
</div></details></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../" data-current-crate="sewup_derive" data-search-index-js="../search-index.js" data-search-js="../search.js"></div>
</body></html>