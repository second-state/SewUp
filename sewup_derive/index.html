<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `sewup_derive` crate."><meta name="keywords" content="rust, rustlang, rust-lang, sewup_derive"><title>sewup_derive - Rust</title><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../ayu.css" disabled ><script id="default-settings" ></script><script src="../storage.js"></script><script src="../crates.js"></script><script defer src="../main.js"></script>
    <noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"><style type="text/css">#crate-search{background-image:url("../down-arrow.svg");}</style></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../sewup_derive/index.html'><div class='logo-container rust-logo'><img src='../rust-logo.png' alt='logo'></div></a><h2 class="location">Crate sewup_derive</h2><div class="block version"><div class="narrow-helper"></div><p>Version 0.1.9</p></div><div class="sidebar-elems"><a id="all-types" href="all.html"><p>See all sewup_derive's items</p></a><div class="block items"><ul><li><a href="#macros">Macros</a></li></ul></div><div id="sidebar-vars" data-name="sewup_derive" data-ty="mod" data-relpath=""></div><script defer src="sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="18" height="18" alt="Pick another theme!" src="../brush.svg"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../settings.html" title="settings"><img width="18" height="18" alt="Change settings" src="../wheel.svg"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Crate <a class="mod" href="#">sewup_derive</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../src/sewup_derive/lib.rs.html#1-1519" title="goto source code">[src]</a></span></h1><h2 id="macros" class="section-header"><a href="#macros">Macros</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.SizedString.html" title="sewup_derive::SizedString macro">SizedString</a></div><div class="item-right docblock-short"><p>help you write the field which storage string no longer than the specific size</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.ewasm_assert_eq.html" title="sewup_derive::ewasm_assert_eq macro">ewasm_assert_eq</a></div><div class="item-right docblock-short"><p>helps you assert output from the handle of a contract with <code>Vec&lt;u8&gt;</code>.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.ewasm_assert_ok.html" title="sewup_derive::ewasm_assert_ok macro">ewasm_assert_ok</a></div><div class="item-right docblock-short"><p>helps you assert your handler without error and returns</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.ewasm_auto_assert_eq.html" title="sewup_derive::ewasm_auto_assert_eq macro">ewasm_auto_assert_eq</a></div><div class="item-right docblock-short"><p>helps you assert return instance from your handler with auto unwrap ewasm_main, namely <code>#[ewasm_main(auto)]</code></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.ewasm_call_only_by.html" title="sewup_derive::ewasm_call_only_by macro">ewasm_call_only_by</a></div><div class="item-right docblock-short"><p>helps you return handler when caller is not in access control list</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.ewasm_err_output.html" title="sewup_derive::ewasm_err_output macro">ewasm_err_output</a></div><div class="item-right docblock-short"><p>helps you to get the binary result of the thiserror,</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.ewasm_fn_sig.html" title="sewup_derive::ewasm_fn_sig macro">ewasm_fn_sig</a></div><div class="item-right docblock-short"><p>helps you get you function signature</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.ewasm_input.html" title="sewup_derive::ewasm_input macro">ewasm_input</a></div><div class="item-right docblock-short"><p>helps you generate the input raw data for specific contract handler</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.ewasm_input_from.html" title="sewup_derive::ewasm_input_from macro">ewasm_input_from</a></div><div class="item-right docblock-short"><p>helps you to get the input data from contract caller</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.ewasm_output_from.html" title="sewup_derive::ewasm_output_from macro">ewasm_output_from</a></div><div class="item-right docblock-short"><p>help you generate the exactly contract output form rust instance</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.ewasm_rusty_assert_ok.html" title="sewup_derive::ewasm_rusty_assert_ok macro">ewasm_rusty_assert_ok</a></div><div class="item-right docblock-short"><p>helps you assert return Ok(()) your handler with rusty ewasm_main, namely <code>#[ewasm_main(rusty)]</code></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.ewasm_rusty_err_output.html" title="sewup_derive::ewasm_rusty_err_output macro">ewasm_rusty_err_output</a></div><div class="item-right docblock-short"><p>helps you assert return Err your handler with rusty ewasm_main, namely <code>#[ewasm_main(rusty)]</code></p>
</div></div></div><h2 id="attributes" class="section-header"><a href="#attributes">Attribute Macros</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="attr" href="attr.ewasm_constructor.html" title="sewup_derive::ewasm_constructor attr">ewasm_constructor</a></div><div class="item-right docblock-short"><p>helps you to build your constructor for the contract</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="attr" href="attr.ewasm_fn.html" title="sewup_derive::ewasm_fn attr">ewasm_fn</a></div><div class="item-right docblock-short"><p>helps you to build your handlers in the contract</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="attr" href="attr.ewasm_lib_fn.html" title="sewup_derive::ewasm_lib_fn attr">ewasm_lib_fn</a></div><div class="item-right docblock-short"><p>helps you to build your handler in other module</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="attr" href="attr.ewasm_main.html" title="sewup_derive::ewasm_main attr">ewasm_main</a></div><div class="item-right docblock-short"><p>helps you setup the main function of a contract</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="attr" href="attr.ewasm_test.html" title="sewup_derive::ewasm_test attr">ewasm_test</a></div><div class="item-right docblock-short"><p>helps you setup the test mododule, and test cases in contract.</p>
</div></div></div><h2 id="derives" class="section-header"><a href="#derives">Derive Macros</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="derive" href="derive.Key.html" title="sewup_derive::Key derive">Key</a></div><div class="item-right docblock-short"><p><code>Key</code> derive help you implement Key trait for the kv feature</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="derive" href="derive.Table.html" title="sewup_derive::Table derive">Table</a></div><div class="item-right docblock-short"><p>provides the handers for CRUD and the Protocol struct to communicate with these handlers.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="derive" href="derive.Value.html" title="sewup_derive::Value derive">Value</a></div><div class="item-right docblock-short"><p><code>Value</code> derive help you implement Value trait for kv feature</p>
</div></div></div></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../" data-current-crate="sewup_derive" data-search-index-js="../search-index.js" data-search-js="../search.js"></div>
</body></html>