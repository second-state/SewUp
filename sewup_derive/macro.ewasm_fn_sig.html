<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="helps you get you function signature"><meta name="keywords" content="rust, rustlang, rust-lang, ewasm_fn_sig"><title>ewasm_fn_sig in sewup_derive - Rust</title><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../ayu.css" disabled ><script id="default-settings" ></script><script src="../storage.js"></script><script src="../crates.js"></script><script defer src="../main.js"></script>
    <noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"><style type="text/css">#crate-search{background-image:url("../down-arrow.svg");}</style></head><body class="rustdoc macro"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../sewup_derive/index.html'><div class='logo-container rust-logo'><img src='../rust-logo.png' alt='logo'></div></a><div class="sidebar-elems"><h2 class="location">Other items in<br><a href="index.html">sewup_derive</a></h2><div id="sidebar-vars" data-name="ewasm_fn_sig" data-ty="macro" data-relpath=""></div><script defer src="sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="18" height="18" alt="Pick another theme!" src="../brush.svg"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../settings.html" title="settings"><img width="18" height="18" alt="Change settings" src="../wheel.svg"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Macro <a href="index.html">sewup_derive</a>::<wbr><a class="macro" href="#">ewasm_fn_sig</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../src/sewup_derive/lib.rs.html#707" title="goto source code">[src]</a></span></h1><div class="docblock item-decl"><pre class="rust macro"><code>ewasm_fn_sig!() { /* proc-macro */ }</code></pre></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>helps you get you function signature</p>
<ol>
<li>provide function name to get function signature from the same namespace,
which function should be decorated with <code>#[ewasm_fn]</code>, for example,
<code>ewasm_fn_sig!(contract_handler)</code></li>
</ol>

<div class='information'><div class='tooltip compile_fail'>ⓘ</div></div><div class="example-wrap"><pre class="rust rust-example-rendered compile_fail"><code><span class="attribute">#[<span class="ident">ewasm_fn</span>]</span>
<span class="kw">fn</span> <span class="ident">decorated_handler</span>(<span class="ident">a</span>: <span class="ident">i32</span>, <span class="ident">b</span>: <span class="ident">String</span>) -&gt; <span class="prelude-ty">Result</span><span class="op">&lt;</span>()<span class="op">&gt;</span> {
    <span class="prelude-val">Ok</span>(())
}

<span class="attribute">#[<span class="ident">ewasm_main</span>]</span>
<span class="kw">fn</span> <span class="ident">main</span>() -&gt; <span class="prelude-ty">Result</span><span class="op">&lt;</span>()<span class="op">&gt;</span> {
    <span class="kw">let</span> <span class="ident">contract</span> <span class="op">=</span> <span class="ident">Contract::new</span>()<span class="question-mark">?</span>;
    <span class="kw">match</span> <span class="ident">contract</span>.<span class="ident">get_function_selector</span>()<span class="question-mark">?</span> {
        <span class="macro">ewasm_fn_sig!</span>(<span class="ident">decorated_handler</span>) =&gt; <span class="macro">ewasm_input_from!</span>(<span class="ident">contract</span> <span class="kw">move</span> <span class="ident">decorated_handler</span>)<span class="question-mark">?</span>,
        <span class="kw">_</span> =&gt; <span class="kw">return</span> <span class="prelude-val">Err</span>(<span class="ident">Error::UnknownHandle</span>.<span class="ident">into</span>()),
    };
    <span class="prelude-val">Ok</span>(())
}</code></pre></div>
<ol start="2">
<li>provide a function name with input parameters then the macro will
calculate the correct functional signature for you.
ex: <code>ewasm_fn_sig!(undecorated_handler( a: i32, b: String ))</code></li>
</ol>

<div class='information'><div class='tooltip compile_fail'>ⓘ</div></div><div class="example-wrap"><pre class="rust rust-example-rendered compile_fail"><code><span class="comment">// some_crate.rs</span>
<span class="kw">pub</span> <span class="kw">fn</span> <span class="ident">decorated_handler</span>(<span class="ident">a</span>: <span class="ident">i32</span>, <span class="ident">b</span>: <span class="ident">String</span>) -&gt; <span class="prelude-ty">Result</span><span class="op">&lt;</span>()<span class="op">&gt;</span> {
    <span class="prelude-val">Ok</span>(())
}</code></pre></div>

<div class='information'><div class='tooltip compile_fail'>ⓘ</div></div><div class="example-wrap"><pre class="rust rust-example-rendered compile_fail"><code><span class="kw">use</span> <span class="ident">some_crate::decorated_handler</span>;

<span class="attribute">#[<span class="ident">ewasm_main</span>]</span>
<span class="kw">fn</span> <span class="ident">main</span>() -&gt; <span class="prelude-ty">Result</span><span class="op">&lt;</span>()<span class="op">&gt;</span> {
    <span class="kw">let</span> <span class="ident">contract</span> <span class="op">=</span> <span class="ident">Contract::new</span>()<span class="question-mark">?</span>;
    <span class="kw">match</span> <span class="ident">contract</span>.<span class="ident">get_function_selector</span>()<span class="question-mark">?</span> {
        <span class="macro">ewasm_fn_sig!</span>(<span class="ident">undecorated_handler</span>(<span class="ident">a</span>: <span class="ident">i32</span>, <span class="ident">b</span>: <span class="ident">String</span>))
            =&gt; <span class="macro">ewasm_input_from!</span>(<span class="ident">contract</span> <span class="kw">move</span> <span class="ident">undecorated_handler</span>)<span class="question-mark">?</span>,
        <span class="kw">_</span> =&gt; <span class="kw">return</span> <span class="prelude-val">Err</span>(<span class="ident">Error::UnknownHandle</span>.<span class="ident">into</span>()),
    };
    <span class="prelude-val">Ok</span>(())
}</code></pre></div>
</div></details></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../" data-current-crate="sewup_derive" data-search-index-js="../search-index.js" data-search-js="../search.js"></div>
</body></html>