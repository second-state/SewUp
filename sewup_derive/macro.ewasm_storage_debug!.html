<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.ewasm_storage_debug.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.ewasm_storage_debug.html">macro.ewasm_storage_debug.html</a>...</p>
    <script>location.replace("macro.ewasm_storage_debug.html" + location.search + location.hash);</script>
</body>
</html>