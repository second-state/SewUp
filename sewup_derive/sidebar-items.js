initSidebarItems({"attr":[["ewasm_constructor","helps you to build your constructor for the contract"],["ewasm_fn","helps you to build your handlers in the contract"],["ewasm_lib_fn","helps you to build your handler in other module"],["ewasm_main","helps you setup the main function of a contract"],["ewasm_test","helps you setup the test mododule, and test cases in contract."]],"derive":[["Key","`Key` derive help you implement Key trait for the kv feature"],["Table","provides the handers for CRUD and the Protocol struct to communicate with these handlers."],["Value","`Value` derive help you implement Value trait for kv feature"]],"macro":[["SizedString","help you write the field which storage string no longer than the specific size"],["ewasm_assert_eq","helps you assert output from the handle of a contract with `Vec<u8>`."],["ewasm_assert_ok","helps you assert your handler without error and returns"],["ewasm_auto_assert_eq","helps you assert return instance from your handler with auto unwrap ewasm_main, namely `#[ewasm_main(auto)]`"],["ewasm_call_only_by","helps you return handler when caller is not in access control list"],["ewasm_err_output","helps you to get the binary result of the thiserror,"],["ewasm_fn_sig","helps you get you function signature"],["ewasm_input","helps you generate the input raw data for specific contract handler"],["ewasm_input_from","helps you to get the input data from contract caller"],["ewasm_output_from","help you generate the exactly contract output form rust instance"],["ewasm_rusty_assert_ok","helps you assert return Ok(()) your handler with rusty ewasm_main, namely `#[ewasm_main(rusty)]`"],["ewasm_rusty_err_output","helps you assert return Err your handler with rusty ewasm_main, namely `#[ewasm_main(rusty)]`"],["ewasm_storage_debug","helps you inspect the storage in test this will print out the detail of storage, you can add description on the storage debug or leave  it empty"]]});