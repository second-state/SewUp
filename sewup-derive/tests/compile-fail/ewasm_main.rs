use sewup_derive::ewasm_main;

#[ewasm_main]
fn main() -> Result<(), ()> {
    Ok(())
}
