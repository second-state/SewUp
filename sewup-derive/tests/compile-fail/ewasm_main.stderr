error: The function wrapped with ewasm_main need to be compiled with wasm32 target
 --> $DIR/ewasm_main.rs:3:1
  |
3 | #[ewasm_main]
  | ^^^^^^^^^^^^^
  |
  = note: this error originates in the attribute macro `ewasm_main` (in Nightly builds, run with -Z macro-backtrace for more info)
